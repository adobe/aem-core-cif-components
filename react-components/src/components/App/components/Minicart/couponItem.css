.root {
    display: grid;
    grid-gap: 0 1rem;
    grid-template-columns: 1fr min-content;
    margin: 0 1.5rem;
}
.couponName {
    line-height: 1.5rem;
}
