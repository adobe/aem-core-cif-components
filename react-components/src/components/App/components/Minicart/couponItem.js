/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useCartState } from './cartContext';
import Kebab from './kebab';
import Section from './section';

import classes from './couponItem.css';

const CouponItem = () => {
    const [{ cart, removeCoupon }] = useCartState();

    const appliedCoupon = cart.applied_coupon ? cart.applied_coupon.code : null;

    return (
        <div className={classes.root}>
            <div className={classes.couponName}>
                Coupon <strong>{appliedCoupon}</strong> applied.
            </div>
            <Kebab>
                <Section text="Remove coupon" onClick={removeCoupon} icon="Trash" />
            </Kebab>
        </div>
    );
};

export default CouponItem;
