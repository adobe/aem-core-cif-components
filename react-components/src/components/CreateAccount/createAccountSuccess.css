.root {
    padding-top: 1.5rem;
}

.body {
    padding: 1.5rem 1rem 1rem;
}

.header {
    font-size: 1.25rem;
    font-weight: 300;
    line-height: 1.25rem;
}

.textBlock {
    margin: 1.5rem 0 1.5rem;
    line-height: 1.875rem;
    font-size: 0.875rem;
}

.signin {
    align-items: center;
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    line-height: 1;
    pointer-events: auto;
    text-align: center;
    white-space: nowrap;
    background: none;
    border: 1px solid rgb(var(--color));
    border-radius: 1.5rem;
    font-size: 0.75rem;
    font-weight: 600;
    height: 2rem;
    min-width: 7.5rem;
    padding: 0 1rem;
    transition-duration: 384ms;
    transition-property: background-color, color;
    transition-timing-function: var(--venia-anim-standard);
    --color: var(--venia-text);
    color: white;
    background-color: rgb(var(--color));
}

.signin:hover {
    color: white;
    background-color: rgb(var(--venia-teal));
}

.actions {
    display: grid;
    justify-items: center;
    margin-top: 1rem;
}
