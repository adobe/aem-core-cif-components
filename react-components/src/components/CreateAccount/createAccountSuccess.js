/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { func } from 'prop-types';
import { useTranslation, Trans } from 'react-i18next';

import Trigger from '../Trigger';
import { useUserContext } from '../../context/UserContext';
import classes from './createAccountSuccess.css';
import { useEffect } from 'react';

const CreateAccountSuccess = props => {
    const [t] = useTranslation('account');
    const { showSignIn } = props;
    const [{ createAccountEmail }, { dispatch }] = useUserContext();

    const handleSignIn = () => {
        dispatch({ type: 'cleanupAccountCreated' });
        showSignIn();
    };

    return (
        <div className={classes.root}>
            <div className={classes.body}>
                <h2 className={classes.header}>
                    {t('account:account-created-title', 'Your account was successfully created')}
                </h2>
                <div className={classes.textBlock}>
                    {/* prettier-ignore */}
                    <Trans i18nKey="account:email-confirmation-info">
                        You will receive a link at <b>{{ email: createAccountEmail }}</b>. Access that link to confirm your email address.
                    </Trans>
                </div>
                <div className={classes.actions}>
                    <Trigger action={handleSignIn}>
                        <span className={classes.signin}>{t('account:sign-in', 'Sign In')}</span>
                    </Trigger>
                </div>
            </div>
        </div>
    );
};

CreateAccountSuccess.propTypes = {
    showSignIn: func.isRequired
};

export default CreateAccountSuccess;
