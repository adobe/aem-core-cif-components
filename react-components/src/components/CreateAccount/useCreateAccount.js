/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useMutation } from '@apollo/client';
import { useUserContext } from '../../context/UserContext';

import MUTATION_CREATE_CUSTOMER from '../../queries/mutation_create_customer.graphql';

export default props => {
    const { showAccountCreated } = props;
    const [{ isSignedIn, createAccountError, inProgress }, { dispatch }] = useUserContext();

    const [createCustomer] = useMutation(MUTATION_CREATE_CUSTOMER);

    const handleCreateAccount = async formValues => {
        dispatch({ type: 'setInProgress' });
        const {
            customer: { email, firstname, lastname },
            password
        } = formValues;
        try {
            await createCustomer({
                variables: { email, password, firstname, lastname }
            });

            dispatch({ type: 'postCreateAccount', accountEmail: email });
            showAccountCreated();
        } catch (error) {
            dispatch({ type: 'createAccountError', error });
        }
    };

    return [{ isSignedIn, createAccountError, inProgress }, { createAccount: handleCreateAccount }];
};
