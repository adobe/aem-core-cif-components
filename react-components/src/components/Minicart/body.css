.root {
    display: grid;
    grid-template-rows: min-content min-content 1fr;
    overflow: auto;
}
