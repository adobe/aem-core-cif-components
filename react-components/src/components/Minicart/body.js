/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useTranslation } from 'react-i18next';

import LoadingIndicator from '../LoadingIndicator';

import EmptyMinicartBody from './emptyMinicartBody';
import classes from './body.css';
import ProductList from './productList';
import CartOptions from './cartOptions';
import Error from './error';
import CouponForm from './couponForm';
import CouponItem from './couponItem';

import { useCartState } from './cartContext';
import DiscountList from './discountList';

const CartLoadingIndicator = () => {
    const [t] = useTranslation('cart');

    return <LoadingIndicator>{t('cart:fetching-data', 'Fetching cart data...')}</LoadingIndicator>;
};

const Body = () => {
    const [{ isEditing, cart, isLoading, errorMessage }] = useCartState();
    const isEmpty = cart && Object.entries(cart).length > 0 ? cart.items.length === 0 : true;

    if (isLoading) {
        return <CartLoadingIndicator />;
    }

    if (errorMessage) {
        return <Error />;
    }

    if (isEmpty) {
        return <EmptyMinicartBody />;
    }

    if (isEditing) {
        return <CartOptions />;
    }

    const cartItems = cart.items;
    const couponFragment = !cart.applied_coupon ? <CouponForm /> : <CouponItem />;
    const discountFragment =
        cart.prices.discounts && cart.prices.discounts.length > 0 ? (
            <DiscountList discounts={cart.prices.discounts} />
        ) : (
            <div />
        );

    return (
        <div className={classes.root}>
            <ProductList cartItems={cartItems} />
            {discountFragment}
            {couponFragment}
        </div>
    );
};

export default Body;
