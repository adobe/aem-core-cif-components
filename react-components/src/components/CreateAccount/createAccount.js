/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { Form } from 'informed';
import { shape, string, func } from 'prop-types';
import Field from '../Field';
import TextInput from '../TextInput';
import Checkbox from '../Checkbox';
import Button from '../Button';
import combine from '../../utils/combineValidators';
import {
    validateEmail,
    isRequired,
    validatePassword,
    validateConfirmPassword,
    hasLengthAtLeast
} from '../../utils/formValidators';
import mergeClasses from '../../utils/mergeClasses';
import defaultClasses from './createAccount.css';
import useCreateAccount from './useCreateAccount';

const LEAD = 'Check out faster, use multiple addresses, track orders and more by creating an account!';

const CreateAccount = props => {
    const { showMyAccount } = props;
    const [{ createAccountError, isSignedIn, isCreatingCustomer }, { createAccount }] = useCreateAccount();

    const handleCreateAccount = formValues => {
        createAccount(formValues);
    };

    const errorMessage = createAccountError ? createAccountError : null;

    if (isSignedIn) {
        showMyAccount();
        return null;
    }
    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <Form className={classes.root} onSubmit={handleCreateAccount}>
            <p className={classes.lead}>{LEAD}</p>
            <Field label="First Name" required={true}>
                <TextInput
                    field="customer.firstname"
                    autoComplete="given-name"
                    validate={isRequired}
                    validateOnBlur
                    aria-label="firstname"
                />
            </Field>
            <Field label="Last Name" required={true}>
                <TextInput
                    field="customer.lastname"
                    autoComplete="family-name"
                    validate={isRequired}
                    validateOnBlur
                    aria-label="lastname"
                />
            </Field>
            <Field label="Email" required={true}>
                <TextInput
                    field="customer.email"
                    autoComplete="email"
                    validate={combine([isRequired, validateEmail])}
                    validateOnBlur
                    aria-label="email"
                />
            </Field>
            <Field label="Password" required={true}>
                <TextInput
                    field="password"
                    type="password"
                    autoComplete="new-password"
                    validate={combine([isRequired, [hasLengthAtLeast, 8], validatePassword])}
                    validateOnBlur
                    aria-label="password"
                />
            </Field>
            <Field label="Confirm Password" required={true}>
                <TextInput
                    field="confirm"
                    type="password"
                    validate={combine([isRequired, validateConfirmPassword])}
                    validateOnBlur
                    aria-label="confirm"
                />
            </Field>
            <div className={classes.subscribe}>
                <Checkbox field="subscribe" label="Subscribe to news and updates" aria-label="subscribe" />
            </div>
            <div className={classes.error}>{errorMessage}</div>
            <div className={classes.actions}>
                <Button disabled={isCreatingCustomer} type="submit" priority="high" aria-label="submit">
                    {'Submit'}
                </Button>
            </div>
        </Form>
    );
};

CreateAccount.propTypes = {
    classes: shape({
        actions: string,
        error: string,
        lead: string,
        root: string,
        subscribe: string
    }),
    showMyAccount: func.isRequired
};

export default CreateAccount;
