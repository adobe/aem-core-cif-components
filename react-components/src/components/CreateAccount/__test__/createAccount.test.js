/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React from 'react';
import { fireEvent, waitForElement } from '@testing-library/react';
import { render } from 'test-utils';
import { useUserContext } from '../../..';
import { CartProvider } from '../../Minicart';
import CreateAccount from '../createAccount';

import MUTATION_CREATE_CUSTOMER from '../../../queries/mutation_create_customer.graphql';

describe('<CreateAccount>', () => {
    beforeEach(() => {
        jest.resetModules();
    });
    it('renders the component', () => {
        const { asFragment } = render(
            <CartProvider initialState={{ cartId: null }} reducerFactory={() => state => state}>
                <CreateAccount showMyAccount={jest.fn()} showAccountCreated={jest.fn()} />
            </CartProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('submits the form properly', async () => {
        const ContextWrapper = () => {
            const [{ createAccountEmail }] = useUserContext();
            let content;
            if (createAccountEmail !== null) {
                content = <div data-testid="success">{createAccountEmail}</div>;
            } else {
                content = <CreateAccount showMyAccount={jest.fn()} showAccountCreated={jest.fn()} />;
            }

            return content;
        };

        const { getByLabelText, getByTestId, container } = render(
            <CartProvider initialState={{ cartId: 'guest123' }} reducerFactory={() => state => state}>
                <ContextWrapper />
            </CartProvider>
        );
        const detailsFromValue = value => {
            return {
                target: {
                    value
                }
            };
        };
        fireEvent.change(getByLabelText('firstname'), detailsFromValue('Iris'));
        fireEvent.change(getByLabelText('lastname'), detailsFromValue('McCoy'));
        fireEvent.change(getByLabelText('email'), detailsFromValue('imccoy@weretail.net'));
        fireEvent.change(getByLabelText('password'), detailsFromValue('Imccoy123'));
        fireEvent.change(getByLabelText('confirm'), detailsFromValue('Imccoy123'));

        fireEvent.click(getByLabelText('submit'));

        const result = await waitForElement(() => {
            return getByTestId('success');
        });

        expect(container.querySelector('.root_error')).toBe(null);
        expect(result).not.toBeUndefined();
        expect(result.textContent).toEqual('imccoy@weretail.net');
    });

    it('handles the account creation error', async () => {
        const mockPerson = {
            email: 'imccoy@weretail.net',
            firstname: 'Iris',
            lastname: 'McCoy'
        };
        const mockPassword = 'Imccoy123';
        const mocks = [
            {
                request: {
                    query: MUTATION_CREATE_CUSTOMER,
                    variables: {
                        firstname: 'Iris',
                        lastname: 'McCoy',
                        email: 'imccoy@weretail.net',
                        password: 'Imccoy123'
                    }
                },
                result: {
                    errors: [
                        {
                            message: 'A customer with the same email address already exists in an associated website.',
                            category: 'graphql-input',
                            locations: [
                                {
                                    line: 2,
                                    column: 5
                                }
                            ],
                            path: ['createCustomer']
                        }
                    ],
                    data: {
                        createCustomer: null
                    }
                }
            }
        ];

        const ContextWrapper = () => {
            const [{ createAccountError }] = useUserContext();
            let content;
            if (createAccountError) {
                content = <div data-testid="success">{createAccountError}</div>;
            } else {
                content = <CreateAccount showMyAccount={jest.fn()} showAccountCreated={jest.fn()} />;
            }

            return content;
        };

        const { getByTestId, getByLabelText } = render(
            <CartProvider initialState={{ cartId: 'guest123' }} reducerFactory={() => state => state}>
                <ContextWrapper />
            </CartProvider>,
            { mocks: mocks }
        );

        const detailsFromValue = value => {
            return {
                target: {
                    value
                }
            };
        };
        fireEvent.change(getByLabelText('firstname'), detailsFromValue(mockPerson.firstname));
        fireEvent.change(getByLabelText('lastname'), detailsFromValue(mockPerson.lastname));
        fireEvent.change(getByLabelText('email'), detailsFromValue(mockPerson.email));
        fireEvent.change(getByLabelText('password'), detailsFromValue(mockPassword));
        fireEvent.change(getByLabelText('confirm'), detailsFromValue(mockPassword));

        fireEvent.click(getByLabelText('submit'));

        const successMessage = await waitForElement(() => {
            return getByTestId('success');
        });

        expect(successMessage).not.toBeUndefined();
        expect(successMessage.textContent).toEqual(
            'A customer with the same email address already exists in an associated website.'
        );
    });
});
