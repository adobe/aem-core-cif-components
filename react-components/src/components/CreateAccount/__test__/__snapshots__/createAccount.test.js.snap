// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateAccount> renders the component 1`] = `
<DocumentFragment>
  <form>
    <p>
      Check out faster, use multiple addresses, track orders and more by creating an account!
    </p>
    <div
      class="root"
    >
      <span
        class="label"
      >
        <span
          class="requiredSymbol"
        />
        First Name
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            aria-label="firstname"
            autocomplete="given-name"
            class="input"
            name="customer.firstname"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        />
      </span>
      <p
        class="root"
      />
    </div>
    <div
      class="root"
    >
      <span
        class="label"
      >
        <span
          class="requiredSymbol"
        />
        Last Name
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            aria-label="lastname"
            autocomplete="family-name"
            class="input"
            name="customer.lastname"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        />
      </span>
      <p
        class="root"
      />
    </div>
    <div
      class="root"
    >
      <span
        class="label"
      >
        <span
          class="requiredSymbol"
        />
        Email
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            aria-label="email"
            autocomplete="email"
            class="input"
            name="customer.email"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        />
      </span>
      <p
        class="root"
      />
    </div>
    <div
      class="root"
    >
      <span
        class="label"
      >
        <span
          class="requiredSymbol"
        />
        Password
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            aria-label="password"
            autocomplete="new-password"
            class="input"
            name="password"
            type="password"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        />
      </span>
      <p
        class="root"
      />
    </div>
    <div
      class="root"
    >
      <span
        class="label"
      >
        <span
          class="requiredSymbol"
        />
        Confirm Password
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            aria-label="confirm"
            class="input"
            name="confirm"
            type="password"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        />
      </span>
      <p
        class="root"
      />
    </div>
    <div>
      <label
        class="root"
      >
        <span
          class="icon"
        />
        <input
          aria-label="subscribe"
          class="input"
          name="subscribe"
          type="checkbox"
        />
        <span
          class="label"
        >
          Subscribe to news and updates
        </span>
      </label>
      <p
        class="root"
      />
    </div>
    <div />
    <div>
      <button
        aria-label="submit"
        class="root_highPriority"
        type="submit"
      >
        <span
          class="content"
        >
          Submit
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
